src/libsyntax/parse/lexer/mod.rs

```diff
@@ -799,7 +799,15 @@ impl<'a> StringReader<'a> {
                             'n' | 'r' | 't' | '\\' | '\'' | '"' | '0' => true,
                             'x' => self.scan_byte_escape(delim, !ascii_only),
                             'u' if self.curr_is('{') => {
-                                self.scan_unicode_escape(delim)
+                                if self.scan_unicode_escape(delim) && ascii_only {
```

Comment:
If `scan_unicode_escape` returns `false`, doesn't this whole branch return `true`? I think that only happens when the unicode escape is not a valid character (e.g. `\u{ffffff}`). Could you make sure that still emits an error?